### PR TITLE
DevBranch to master

### PR DESCRIPTION
**Pull request by SergeyBadyrov**

### CODE REVIEW

File: HomeWork4/Task3/Program.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            var name = Console.ReadLine();
+
+            Console.WriteLine(name.SayHello());
+            Console.WriteLine(name.SayGoodbye());
+
+            Console.ReadKey();
+        }
+    }
+
+    static class Greetings
```
Comment (NTcvilikh) [inline HomeWork4/Task3/Program.cs]: It's good practice to create every class in separate file

File: HomeWork4/Task3/Program.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        public static string SayHello(this string str)
+        {
+            return "Hello, " + str + "!";
```
Comment (NTcvilikh) [inline HomeWork4/Task3/Program.cs]: It would be great to define the behavior for null or empty string

File: HomeWork4/Task2/Program.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+namespace Task2
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            Stopwatch sw = new Stopwatch();
+
+            //Int array for checking the speed of the algoritm
+            int[] array = new int[10000000];
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: should be declared as constant

File: HomeWork4/Task2/Program.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            Random rand = new Random();
+            for (int i = 0; i < 10000000; i++)
+            {
+                array[i] = rand.Next(0, 999999);
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: should be declared as an another constant

File: HomeWork4/Task2/Program.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            }
+
+            sw.Start();
+            var resultEuckid = GCDEuclid.GetGreatestCommonDivisor(array);
+            sw.Stop();
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: no need

File: HomeWork4/Task2/Program.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            TimeSpan ts = sw.Elapsed;
+            var timeEuclid = FormatTime(ts);
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: FormatTime(sw.Elapsed) is enough

File: HomeWork4/Task2/Program.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            Console.WriteLine("Results of Euclid algorithm. GCD: {0}; time: {1}", resultEuckid, timeEuclid);
+
+            sw.Reset();
+
+            sw.Start();
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: no need

File: HomeWork4/Task2/Program.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            Stopwatch sw = new Stopwatch();
+
+            //Int array for checking the speed of the algoritm
+            int[] array = new int[10000000];
+            Random rand = new Random();
+            for (int i = 0; i < 10000000; i++)
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: array.Length would be better

File: HomeWork4/Task2/Program.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            Console.WriteLine("Results of Euclid algorithm. GCD: {0}; time: {1}", resultEuckid, timeEuclid);
+
+            sw.Reset();
+
+            sw.Start();
+            var resultStein = GCDStein.GetGreatestCommonDivisor(array);
+            sw.Stop();
+            ts = sw.Elapsed;
+            var timeStein = FormatTime(ts);
+            Console.WriteLine("Results of Stein algorithm. GCD: {0}; time: {1}", resultEuckid, timeStein);
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: resultStein is expected

File: HomeWork4/Task2/Program.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+            Console.ReadKey();
+        }
+
+        private static string FormatTime(TimeSpan ts)
+        {
+            return String.Format("{0:00}:{1:00}:{2:00}.{3:00}", ts.Hours, ts.Minutes, ts.Seconds, ts.Milliseconds / 10);
```
Comment (NTcvilikh) [inline HomeWork4/Task2/Program.cs]: can be simplified (if you use MS VS 2015 or greater)
$"{ts.Hours:00}:{ts.Minutes:00}:{ts.Seconds:00}.{ts.Milliseconds / 10:00}";

also can be used default converter
return ts.ToString("c");
or custom
return ts.ToString("mm':'ss'.'ffff");

File: HomeWork4/Task2/GCDEuclid.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace Task2
+{
+    static class GCDEuclid
```
Comment (NTcvilikh) [inline HomeWork4/Task2/GCDEuclid.cs]: static classes is bad practice. static can be used for extensions and could be used for singletons earlier.

File: HomeWork4/Task2/GCDEuclid.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        public static int GetGreatestCommonDivisor(params int[] numbers)
```
Comment (NTcvilikh) [inline HomeWork4/Task2/GCDEuclid.cs]: both GCDEuclid & GCDStein dothe same thing: return the greatest divisor. 
the interface can be introduced here.

File: HomeWork4/Task2/GCDEuclid.cs [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            int gcd = numbers[0];
```
Comment (NTcvilikh) [inline HomeWork4/Task2/GCDEuclid.cs]: null ref can appear here.
always check the input data in public methods.

File: HomeWork4/Task2/GCDEuclid.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+            for (int i = 0; i < (numbers.Length - 1); i++)
```
Comment (NTcvilikh) [inline HomeWork4/Task2/GCDEuclid.cs]: no need in ()

File: HomeWork4/Task2/GCDEuclid.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+            {
+                gcd = EuclidsMethod(gcd, numbers[i + 1]);
```
Comment (NTcvilikh) [inline HomeWork4/Task2/GCDEuclid.cs]: indexOutOfRange can appear here

File: HomeWork4/Task2/GCDStein.cs [hunk 16]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace Task2
+{
+    static class GCDStein
```
Comment (NTcvilikh) [inline HomeWork4/Task2/GCDStein.cs]: same notes as for GCDEuclid class

File: HomeWork4/Task2/GCDStein.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+            if (number1 == number2) return number1;                            
+            if (number1 == 1 || number2 == 1) return 1;
+
+            if ((number1 & 1) == 0)
+            {
+                return ((number2 & 1) == 0) ? SteinsMethod(number1 >> 1, number2 >> 1) << 1 : SteinsMethod(number1 >> 1, number2);
+            }
+            else
+            {
+                if ((number2 & 1) == 0)
```
Comment (NTcvilikh) [inline HomeWork4/Task2/GCDStein.cs]: else if can be united. pls avoid unnecessarily nesting

File: HomeWork4/Task1/Program.cs [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+            var output = ConvertInIEEE754(number);
+            Console.WriteLine(output);
+
+            Console.ReadKey();
+        }
+
+        private static string ConvertInIEEE754(double number)
+        {
+            Int64 a = BitConverter.DoubleToInt64Bits(number);
```
Comment (NTcvilikh) [inline HomeWork4/Task1/Program.cs]: Not enough. See http://weitz.de/ieee/
[TestCase(255.16, "01000000011011111110010100011110")]
[TestCase(-20, "11000000001101000000000000000000")]

File: HomeWork4/Task1/Program.cs [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+            Console.WriteLine("Введите число для конвертации: ");
+            var number = Convert.ToDouble(Console.ReadLine());
+
+            var output = ConvertInIEEE754(number);
+            Console.WriteLine(output);
+
+            Console.ReadKey();
+        }
+
+        private static string ConvertInIEEE754(double number)
```
Comment (NTcvilikh) [inline HomeWork4/Task1/Program.cs]: it's good practice to place extensions is a separate static class

File: HomeWork3/Task1/Program.cs [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+            Console.WriteLine("Enter the natural degree");
+            var degree = Convert.ToInt32(Console.ReadLine());
+
+            Console.WriteLine("Enter the real accuracy");
+            var accuracy = Convert.ToDouble(Console.ReadLine());
+
+            Console.WriteLine("Result of the Math.Pow method: " + Math.Pow(number, 1.0 / degree));
+
+            var root = Newton.NewtonMethod(number, degree, accuracy);
+            Console.WriteLine("Result of the custom Newton's method: " + root.ToString());
```
Comment (NTcvilikh) [inline HomeWork3/Task1/Program.cs]: no need in .ToString

File: HomeWork3/Task1/Newton.cs [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Task1
+{
+    static class Newton
+    {
+        public static double NewtonMethod(double number, int degree, double eps)
+        {
+            double x0 = number / (double)degree;
```
Comment (NTcvilikh) [inline HomeWork3/Task1/Newton.cs]: (double)degree can be stored in local var

File: HomeWork3/Task1/Newton.cs [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Task1
+{
+    static class Newton
+    {
+        public static double NewtonMethod(double number, int degree, double eps)
```
Comment (NTcvilikh) [inline HomeWork3/Task1/Newton.cs]: input data isn't checked. ex can appear.

File: HomeWork3/Task2/Context.cs [hunk 23]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Task2
```
Comment (NTcvilikh) [inline HomeWork3/Task2/Context.cs]: not pure strategy. i'll discuss with you and Mark at holidays after you fix all other issues in all your tasks.

File: HomeWork3/Task2/Program.cs [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            int[,] array = { { 4, 7, 8 }, { 2, 9, 5 }, { 3, 8, 1 }, { 6, 7, 3 } };
+            MatrixOperations.PrintMatrix(array);
+
+            Context context = new Context();
+            Console.WriteLine("Choose the operation number - from 1 to 6");
+            Operation operation = (Operation) Convert.ToInt32(Console.ReadLine());
```
Comment (NTcvilikh) [inline HomeWork3/Task2/Program.cs]: can be optimized

File: HomeWork3/Task2/Program.cs [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+            switch (operation)
+            {
+                case (Operation)1:
```
Comment (NTcvilikh) [inline HomeWork3/Task2/Program.cs]: case Operation.OrderByRowSumInc will work

File: HomeWork2/MainPage/Task5Class.cs [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MainPage
+{
+    class Task5Class
+    {
+        public void FilterLucky(List<int> list)
+        {
+            foreach(int item in list.ToArray())
```
Comment (NTcvilikh) [inline HomeWork2/MainPage/Task5Class.cs]: can be optimized. use list.RemoveAll. it's List's method, not LINQ which you'll learn in the future.

File: HomeWork2/MainPage/Task4Class.cs [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Tasks;
+
+namespace MainPage
+{
+    class Task4Class
+    {
+        public string AlphabeticalConcatenation(string str1, string str2)
+        {
+            if (string.IsNullOrWhiteSpace(str1) || string.IsNullOrWhiteSpace(str2))
+                return "Both strings are empty!";
```
Comment (NTcvilikh) [inline HomeWork2/MainPage/Task4Class.cs]: nope. one of

File: HomeWork2/MainPage/Task4Class.cs [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+            if (string.IsNullOrWhiteSpace(str1) && string.IsNullOrWhiteSpace(str2))
+                return "One of the strings is empty!";
```
Comment (NTcvilikh) [inline HomeWork2/MainPage/Task4Class.cs]: nope. both

File: HomeWork2/MainPage/Task4Class.cs [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+
+            string result = $"{str1 + str2}";
+            char[] resultChar = new char[result.Length];
+            for (int i = 0; i < resultChar.Length; i++)
+            {
+                resultChar[i] = result[i];
```
Comment (NTcvilikh) [inline HomeWork2/MainPage/Task4Class.cs]: use str.ToCharArray instead

File: HomeWork2/MainPage/Task4Class.cs [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+            }
+
+            resultChar = Sort(resultChar);
```
Comment (NTcvilikh) [inline HomeWork2/MainPage/Task4Class.cs]: wrong naming. it's not a result and not a char. "var chars" 'll be ok

File: HomeWork2/MainPage/Task4Class.cs [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+
+            result = "" + resultChar[0];
+            for (int i = 1; i < resultChar.Length; i++)
+            {
+                if (resultChar[i] == resultChar[i - 1]) { continue; }
```
Comment (NTcvilikh) [inline HomeWork2/MainPage/Task4Class.cs]: inverted 'if' would be more beautiful 
                if (resultChar[i] != resultChar[i - 1])
                    result += resultChar[i];

File: HomeWork2/MainPage/Task1Class.cs [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+            int number1RightPart;
+            int result;
+
+            //For saving initial lenghts of number1 and number2
+            int number1Length = Convert.ToString(number1, 2).Length;
+            int number2Length = Convert.ToString(number2, 2).Length;
+
+            //extraBit shows should we enlarge or decrease the length of the final number
+            int extraBit;
+            if ((j - i + 1) < number2Length || (j - i + 1) > number2Length)
```
Comment (NTcvilikh) [inline HomeWork2/MainPage/Task1Class.cs]: j - i + 1 can be stored in a var

Comment (NTcvilikh) [inline HomeWork2/MainPage/Task1Class.cs]: ternary operator ?: can be used

File: .gitignore [hunk 33]
```diff
@@ -0,0 +1,288 @@
+## Ignore Visual Studio temporary files, build results, and
```
Comment (NTcvilikh) [inline .gitignore]: stub comment

File: HomeWork2/Task1/Task1Class.cs [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+                int number1RightPart;
+                int result;
+
+                //For saving initial lenghts of number1 and number2
+                int number1Length = Convert.ToString(number1, 2).Length;
+                int number2Length = Convert.ToString(number2, 2).Length;
+
+                //extraBit shows should we enlarge or decrease the length of the final number
+                int extraBit;
+                if ((j - i + 1) < number2Length || (j - i + 1) > number2Length)
```
[the remark this answers is not in the document]
Comment (SergeyBadyrov) [reply HomeWork2/Task1/Task1Class.cs]: Here is important to show that the 2nd number is not equal to the number of bits to insert
And if it is not, get the difference between their lengths

File: HomeWork4/Task3/Greetings.cs [hunk 35]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Task3
+{
+    public static class Greetings
```
[the remark this answers is not in the document]
Comment (SergeyBadyrov) [reply HomeWork4/Task3/Greetings.cs]: sure, but for HW4 task was: Develop unit tests (NUnit and (or) MS Unit Test) for Tasks 1, 2 Module 4.

File: HomeWork4/Task3/Greetings.cs [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        public static string SayHello(this string str)
```
[the remark this answers is not in the document]
Comment (SergeyBadyrov) [reply HomeWork4/Task3/Greetings.cs]: was already done

**Pull request merged**